### PR TITLE
Improve FleetWnd split fleet

### PR DESCRIPTION
**Pull request by LGM-Doyle**
This PR improves the run time of `FleetWnd`split fleet.  

It will also improve update times for any maps which contain fleets.  

It does not change any game rules, or UI elements.  I think it is suitable for inclusion in 0.4.6.

On my test example merging all the fleets in a fleet button with many fleets and then splitting them.  The splitting took 19 seconds before this PR and 3 seconds after.  

You can compare the results yourself with any save with a large fleet. The PR is save compatible.  

The first commit https://github.com/freeorion/freeorion/commit/efa00c3441b7d24286baaf9df1e052bc0dd8bea4 contains the `FleetWnd::SplitFleet` `ScopedTimer` for split fleet.  If you checkout https://github.com/freeorion/freeorion/commit/efa00c3441b7d24286baaf9df1e052bc0dd8bea4 and split a big fleet, the time will be in the `freeorion.log` as `FleetWnd::SplitFleet time: xxxx.xx`.  You can compare the results with the whole PR.  The changes are large enough to be noticeable without the timer.

The PR makes several changes in `MapWnd` that cumulatively produce the improvement.
- it creates separate handlers for fleet insert and remove events.
- it defers all work from multiple fleet insert/remove events during a single render interval to a single function `DeferredRefreshFleetButtons()` run only if needed at the start of `Render()`
- it removes some redundancy in fleet movement line drawing.
- it only examines fleets instead of all objects when considering fleets to draw.
- it uses unordered sets and maps for unordered sets and maps of fleet related objects.

It makes some code quality changes, by reducing repeated code.

### CODE REVIEW

**Comment by MatGB** [comment]
For reviewers: I've tested this and it works fine on my system, with a noticeable improvement in performance, I cannot comment on the code itself but would recommend inclusion ASAP if possible.

File: UI/MapWnd.cpp [hunk 1]
```diff
@@ -4268,36 +4263,39 @@ void MapWnd::DoFleetButtonsLayout() {
...
             ++n;
             (*button_it)->MoveTo(ul + icon_ul);
         }
     }
 
     // position moving fleet buttons
-    for (std::set<FleetButton*>::iterator it = m_moving_fleet_buttons.begin(); it != m_moving_fleet_buttons.end(); ++it) {
-        FleetButton* fb = *it;
+    for (boost::unordered_map<std::pair<double, double>, boost::unordered_set<FleetButton*> >::iterator pos_it = m_moving_fleet_buttons.begin();
+         pos_it != m_moving_fleet_buttons.end(); ++pos_it) {
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: { to next line

File: UI/MapWnd.cpp [hunk 2]
```diff
@@ -4324,292 +4322,222 @@ std::pair<double, double> MapWnd::MovingFleetMapPositionOnLane(TemporaryPtr<cons
...
-        // skip known destroyed and stale info objects
-        if (this_client_known_destroyed_objects.find(object_id) != this_client_known_destroyed_objects.end())
-            continue;
-        if (this_client_stale_object_info.find(object_id) != this_client_stale_object_info.end())
-            continue;
+        if (fleet
+            && !fleet->Empty()
+            && (known_destroyed_objects.count(object_id) == 0)
+            && (stale_object_info.count(object_id) == 0))
+            return fleet;
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp:4344]: need { } around nested code after multi-like if clause

File: UI/MapWnd.cpp [hunk 3]
```diff
@@ -4324,292 +4322,222 @@ std::pair<double, double> MapWnd::MovingFleetMapPositionOnLane(TemporaryPtr<cons
...
+        return TemporaryPtr<const Fleet>();
+    }
+
+    /** If the \p fleet has orders and is departing from a valid system, return the system*/
+    TemporaryPtr<const System> IsDepartingFromSystem(const TemporaryPtr<const Fleet>& fleet) {
+
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: unnecessary blank line?

File: UI/MapWnd.cpp [hunk 4]
```diff
@@ -4324,292 +4322,222 @@ std::pair<double, double> MapWnd::MovingFleetMapPositionOnLane(TemporaryPtr<cons
...
+        if (fleet->FinalDestinationID() != INVALID_OBJECT_ID
+            && !fleet->TravelRoute().empty()
+            && fleet->SystemID() != INVALID_OBJECT_ID) {
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: { to next line

File: UI/MapWnd.cpp [hunk 5]
```diff
@@ -4324,292 +4322,222 @@ std::pair<double, double> MapWnd::MovingFleetMapPositionOnLane(TemporaryPtr<cons
...
+                          << " of a departing fleet named " << fleet->Name();
+        }
+        return TemporaryPtr<const System>();
+    }
+
+    /** If the \p fleet is stationary in a valid system, return the system*/
+    TemporaryPtr<const System> IsStationaryInSystem(const TemporaryPtr<const Fleet>& fleet) {
+        if ((fleet->FinalDestinationID() == INVALID_OBJECT_ID
+             || fleet->TravelRoute().empty())
+            && fleet->SystemID() != INVALID_OBJECT_ID) {
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: ...

File: UI/MapWnd.cpp [hunk 6]
```diff
@@ -4324,292 +4322,222 @@ std::pair<double, double> MapWnd::MovingFleetMapPositionOnLane(TemporaryPtr<cons
...
-        TemporaryPtr<const UniverseObject> obj = *it;
-        int object_id = obj->ID();
+    SystemXEmpireToFleetsMap   departing_fleets;
+    SystemXEmpireToFleetsMap   stationary_fleets;
+    LocationXEmpireToFleetsMap moving_fleets;
 
-        if (verbose_logging)
-            DebugLogger() << "stationary fleet id: " << object_id;
+    for (std::map<int, TemporaryPtr<UniverseObject> >::const_iterator candidate_it = Objects().ExistingFleetsBegin();
+         candidate_it != Objects().ExistingFleetsEnd(); ++candidate_it) {
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: ...

File: UI/MapWnd.cpp [hunk 7]
```diff
@@ -4324,292 +4322,222 @@ std::pair<double, double> MapWnd::MovingFleetMapPositionOnLane(TemporaryPtr<cons
...
-    // moving along starlanes: "moving fleets"
-    std::map<std::pair<double, double>, std::map<int, std::vector<TemporaryPtr<const Fleet> > > > moving_fleets;
-    std::vector<TemporaryPtr<const UniverseObject> > moving_fleet_objects = objects.FindObjects(MovingFleetVisitor());
-    for (std::vector<TemporaryPtr<const UniverseObject> >::iterator it = moving_fleet_objects.begin();
-         it != moving_fleet_objects.end(); ++it)
+template <typename K>
+void MapWnd::CreateFleetButtonsOfType (
+    boost::unordered_map<K, boost::unordered_set<FleetButton*> >& type_fleet_buttons,
+    const boost::unordered_map<std::pair<K, int>, std::vector<int> > &fleets_map,
+    const FleetButton::SizeType & fleet_button_size) {
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: ...

File: UI/MapWnd.cpp [hunk 8]
```diff
@@ -4324,292 +4322,222 @@ std::pair<double, double> MapWnd::MovingFleetMapPositionOnLane(TemporaryPtr<cons
...
 
-    // clear old fleet buttons
+void MapWnd::DeleteFleetButtons() {
     m_fleet_buttons.clear();            // duplicates pointers in following containers
 
-    for (std::map<int, std::set<FleetButton*> >::iterator it = m_stationary_fleet_buttons.begin(); it != m_stationary_fleet_buttons.end(); ++it)
-        for (std::set<FleetButton*>::iterator set_it = it->second.begin(); set_it != it->second.end(); ++set_it)
+    for (boost::unordered_map<int, boost::unordered_set<FleetButton*> >::iterator it = m_stationary_fleet_buttons.begin();
+         it != m_stationary_fleet_buttons.end(); ++it)
+        for (boost::unordered_set<FleetButton*>::iterator set_it = it->second.begin(); set_it != it->second.end(); ++set_it)
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp:4492]: need { after multi-line for

File: UI/MapWnd.cpp [hunk 9]
```diff
@@ -4324,292 +4322,222 @@ std::pair<double, double> MapWnd::MovingFleetMapPositionOnLane(TemporaryPtr<cons
...
             delete *set_it;
     m_stationary_fleet_buttons.clear();
 
-    for (std::map<int, std::set<FleetButton*> >::iterator it = m_departing_fleet_buttons.begin(); it != m_departing_fleet_buttons.end(); ++it)
-        for (std::set<FleetButton*>::iterator set_it = it->second.begin(); set_it != it->second.end(); ++set_it)
+    for (boost::unordered_map<int, boost::unordered_set<FleetButton*> >::iterator it = m_departing_fleet_buttons.begin();
+         it != m_departing_fleet_buttons.end(); ++it)
+        for (boost::unordered_set<FleetButton*>::iterator set_it = it->second.begin(); set_it != it->second.end(); ++set_it)
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp:4500]: {

File: UI/MapWnd.cpp [hunk 10]
```diff
@@ -4324,292 +4322,222 @@ std::pair<double, double> MapWnd::MovingFleetMapPositionOnLane(TemporaryPtr<cons
...
             delete *set_it;
     m_departing_fleet_buttons.clear();
 
-    for (std::set<FleetButton*>::iterator set_it = m_moving_fleet_buttons.begin(); set_it != m_moving_fleet_buttons.end(); ++set_it)
-        delete *set_it;
+    for (boost::unordered_map<std::pair<double, double>, boost::unordered_set<FleetButton*> >::iterator
+             it = m_moving_fleet_buttons.begin(); it != m_moving_fleet_buttons.end(); ++it)
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: make indenting match start of boost:: on previous line
need { } after multi-line for

File: UI/MapWnd.cpp [hunk 11]
```diff
@@ -5261,27 +5185,29 @@ void MapWnd::RefreshFleetButtonSelectionIndicators() {
     //std::cout << "MapWnd::RefreshFleetButtonSelectionIndicators()" << std::endl;
 
     // clear old selection indicators
-    for (std::map<int, std::set<FleetButton*> >::iterator it = m_stationary_fleet_buttons.begin(); it != m_stationary_fleet_buttons.end(); ++it) {
-        std::set<FleetButton*>& set = it->second;
-        for (std::set<FleetButton*>::iterator button_it = set.begin(); button_it != set.end(); ++button_it)
+    for (boost::unordered_map<int, boost::unordered_set<FleetButton*> >::iterator it = m_stationary_fleet_buttons.begin(); it != m_stationary_fleet_buttons.end(); ++it) {
+        boost::unordered_set<FleetButton*>& set = it->second;
+        for (boost::unordered_set<FleetButton*>::iterator button_it = set.begin(); button_it != set.end(); ++button_it)
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp]: {

File: UI/MapWnd.cpp [hunk 12]
```diff
@@ -5261,27 +5185,29 @@ void MapWnd::RefreshFleetButtonSelectionIndicators() {
...
             (*button_it)->SetSelected(false);
     }
 
-    for (std::map<int, std::set<FleetButton*> >::iterator it = m_departing_fleet_buttons.begin(); it != m_departing_fleet_buttons.end(); ++it) {
-        std::set<FleetButton*>& set = it->second;
-        for (std::set<FleetButton*>::iterator button_it = set.begin(); button_it != set.end(); ++button_it)
+    for (boost::unordered_map<int, boost::unordered_set<FleetButton*> >::iterator it = m_departing_fleet_buttons.begin(); it != m_departing_fleet_buttons.end(); ++it) {
+        boost::unordered_set<FleetButton*>& set = it->second;
+        for (boost::unordered_set<FleetButton*>::iterator button_it = set.begin(); button_it != set.end(); ++button_it)
```
Comment (geoffthemedio) [inline UI/MapWnd.cpp:5210]: {

**Comment by LGM-Doyle** [comment]
@geoffthemedio thanks for the prompt review.

I've made all the requested changes.

**Pull request merged**